lib/apiservers/engine/backends/container.go

```diff
@@ -882,198 +877,174 @@ func IsContainerRunning(containerID string) (bool, error) {
-func attachStreams(name string, tty, stdinOnce bool, clStdin io.ReadCloser, clStdout, clStderr io.Writer, keys []byte) error {
-	var wg sync.WaitGroup
+func attachStreams(ctx context.Context, vc *viccontainer.VicContainer, clStdin io.ReadCloser, clStdout, clStderr io.Writer, ca *backend.ContainerAttachConfig) error {
+	defer clStdin.Close()
 
+	// Cancel will close the child connections.
+	ctx, cancel := context.WithCancel(ctx)
+
+	var wg sync.WaitGroup
 	errors := make(chan error, 3)
+
 	//FIXME: Swagger will timeout on us.  We need to either have an infinite timeout or the timeout should
 	// start after some inactivity?
-	inContext, inCancel := context.WithTimeout(context.Background(), attachRequestTimeout)
-	outContext, outCancel := context.WithTimeout(context.Background(), attachRequestTimeout)
-	errContext, errCancel := context.WithTimeout(context.Background(), attachRequestTimeout)
-
-	if clStdin != nil {
-		wg.Add(1)
-		// Pipe for stdin so we can interject and watch the input streams for detach keys.
-		stdinReader, stdinWriter := io.Pipe()
-
-		defer clStdin.Close()
-		defer stdinReader.Close()
-
-		// For stdin, we only have a timeout for connection.  There can be a long duration before
-		// the first entry so there is no timeout for attempt.
-		plClient, transport := createNewAttachClientWithTimeouts(attachConnectTimeout, 0, 0)
-		defer transport.Close()
-
-		// Swagger wants an io.reader so give it the reader pipe.  Also, the swagger call
-		// to set the stdin is synchronous so we need to run in a goroutine
-		go func() {
-			setStdinParams := interaction.NewContainerSetStdinParamsWithContext(inContext).WithID(name)
-			setStdinParams = setStdinParams.WithRawStream(stdinReader)
-			_, err := plClient.Interaction.ContainerSetStdin(setStdinParams)
 
-			if err != nil {
-				if netErr, ok := err.(net.Error); ok {
-					if netErr.Timeout() {
-						log.Errorf("Net timeout for container stdin")
-					} else {
-						log.Errorf("Net error for container stdin: %s", netErr.Error())
-					}
-				}
-			}
-		}()
+	// For stdin, we only have a timeout for connection.  There can be a long duration before
+	// the first entry so there is no timeout for attempt.
+	plClient, transport := createNewAttachClientWithTimeouts(attachConnectTimeout, 0, 0)
+	defer transport.Close()
 
-		// Copy the stdin from the CLI and write to a pipe.  We need to do this so we can
-		// watch the stdin stream for the detach keys.
+	if ca.UseStdin {
+		wg.Add(1)
 		go func() {
-			var err error
 			defer wg.Done()
-
-			if tty {
-				_, err = copyEscapable(stdinWriter, clStdin, keys)
+			err := copyStdIn(ctx, plClient, vc, clStdin, ca.DetachKeys)
+			if err != nil {
+				log.Errorf("container attach: stdin (%s): %s", vc.ContainerID, err.Error())
 			} else {
-				_, err = io.Copy(stdinWriter, clStdin)
+				log.Infof("container attach: stdin (%s) done: %s", vc.ContainerID)
 			}
 
-			if err != nil {
-				log.Errorf(err.Error())
-			}
+			cancel()
```

Comment:
There was a prior comment associated with this block - it's been moved to copyStdin so nevermind.